### ISSUE TITLE
Setting sync is too slow

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#96373]
re #96226

* have two machines
* change theme on machine A
* wait a minute on machine B - nothing happens
* change theme on machine B - immediate conflict

**Comment by sandy081** [comment]
> wait a minute on machine B - nothing happens

We sync in an interval of every 5 minutes and also during following 

- Window is focused
- Settings or Keybindings or Extensions are opened or changed

We do not have server push mechanism to know if something has changed for a given account. 

Any other suggestions?

**Comment by jrieken** [comment]
> We do not have server push mechanism to know if something has changed for a given account.

Yes - have that

**Comment by sandy081** [comment]
IMO Server push makes sync more instant but is not trivial. @Tyriar any idea if it is possible?

We are already aggressive in triggering sync which might not be good enough for this scenario. But I think this is a test/demo scenario and not real time. It would be great to know a real time scenario that requires more instant sync.

**Comment by jrieken** [comment]
Well, there is no "sync now"-command which means I cannot really see sync at work. I just have to wait or know what command triggers sync as a side-effect. That makes this black-box'ish

I also wonder if a push model would be better then frequent pull because I don't think user change sync-able items that often. Now, you have a request at least every 5mins. Assuming we average at 1000 machines over 24hrs that makes 288000 (12*24*1000) requests. I doubt that users change settings, keybindings that often and I would assume that most request are answered with "nothing to do". So, how about a combination of push and pull

* client pull on request (Sync now action)
* client pull rarely (on start, once a day?)
* server pushes messages that signal the client should pull now

Already today we should know often settings et al change and how often the server returns nothing but my gut feeling tells me that this would reduce server load by a lot. So, apart from the obvious user-observable benefits this might pay off (for real)

**Comment by sandy081** [comment]
Yes, I do agree that pull model reduces the server load but this requires active/persistent connections with the server. Active connections are unnecessary overhead if the client is not active. 

In push model we are reducing the overload on server by using manifests and caching last sync data mechanism @Tyriar mentioned before. See - https://github.com/microsoft/vscode/blob/master/src/vs/platform/userDataSync/common/abstractSynchronizer.ts#L248. So out of `12241000` requests, only those which are not hitting the cache are relevant (@Tyriar any estimates here?)

We need to evaluate the overhead between 1000 active connections or N active requests. I do not know which is better from server point of view.

Re sync now action, I do not think this is necessary if Sync is instant enough - https://github.com/microsoft/vscode/issues/88249 (under discussion already)
